bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/CommunicationManager.java

```diff
@@ -353,6 +373,44 @@ private void receiveUpdate(ItemStateEvent updateEvent) {
         return null;
     }
 
+    private boolean hasDimension(Item item, @Nullable String acceptedItemType) {
+        return (item instanceof NumberItem && ((NumberItem) item).getDimension() != null)
+                || getDimension(acceptedItemType) != null;
+    }
+
+    private @Nullable QuantityType<?> convertToQuantityType(DecimalType originalType, Item item,
+            @Nullable String acceptedItemType) {
+        NumberItem numberItem = (NumberItem) item;
+
+        // DecimalType command sent via a NumberItem with dimension:
+        Class<? extends Quantity<?>> dimension = numberItem.getDimension();
+
+        if (dimension == null) {
+            // DecimalType command sent via a plain NumberItem w/o dimension.
+            // We try to guess the correct unit from the channel-type's expected item dimension
+            // or from the item's state description.
+            dimension = getDimension(acceptedItemType);
```

Comment:
Regarding the "try to guess", do you know of any possible scenarios where the guessing would fail? Will the guessing be succesfull when both the channel type and item state description use `%item%`?

Reply:
According to what I can see (and @htreu please correct if wrong) it'll fall back to the default unit which the UnitProvider returns for that dimension. Which IMHO makes sense, as it is the "most likely" unit. Still - it's just a guess.

Reply:
AFAIK the UnitProvider does not provide units for dimensions like Power. So will the guessing fail for the Power dimension when the channel and item both use `%unit%`? Or will it use the current unit of the item state?

Reply:
I did some testing with a local build of this PR and the Nest Binding with HABPanel similar to this [community post](https://community.openhab.org/t/solved-rest-api-nest-temperature-item-problem/45817/15?u=wborn) with the following items:

```php
Number:Temperature Thermostat_SP_C "Set Point [%.1f °C]" { channel="nest:thermostat:eca8c465:PuNbMpQGIY3hk7Nf8qJh6secrx8nUoq_:set_point" }
Number:Temperature Thermostat_SP_F "Set Point [%.1f °F]" { channel="nest:thermostat:eca8c465:PuNbMpQGIY3hk7Nf8qJh6secrx8nUoq_:set_point" }
Number:Temperature Thermostat_SP_K "Set Point [%.1f K]"  { channel="nest:thermostat:eca8c465:PuNbMpQGIY3hk7Nf8qJh6secrx8nUoq_:set_point" }
Number:Temperature Thermostat_SP_U "Set Point [%.1f %unit%]" { channel="nest:thermostat:eca8c465:PuNbMpQGIY3hk7Nf8qJh6secrx8nUoq_:set_point" }
```

I've added some logging to the binding to show what commands are received. Commands on the first three items now get a proper unit :+1: :

```
22:47:26.620 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'Thermostat_SP_C' received command 21
22:47:26.621 [INFO ] [ng.nest.handler.NestThermostatHandler] - >>> Received QuantityType command: 21 ℃
22:47:26.621 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_C changed from 20.0 ℃ to 21.0 ℃
22:47:27.252 [INFO ] [smarthome.event.ItemStateChangedEvent] - nest_thermostat_eca8c465_PuNbMpQGIY3hk7Nf8qJh6secrx8nUoq__set_point changed from 20.0 ℃ to 21.0 ℃
22:47:27.253 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_K changed from 293.15 K to 294.15 K
22:47:27.253 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_F changed from 68.00 °F to 69.80 °F
22:47:27.254 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_U changed from 287.0 ℃ to 21.0 ℃
22:47:31.756 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'Thermostat_SP_F' received command 71
22:47:31.757 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_F changed from 69.80 °F to 71.0 °F
22:47:31.757 [INFO ] [ng.nest.handler.NestThermostatHandler] - >>> Received QuantityType command: 71 °F
22:47:32.372 [INFO ] [smarthome.event.ItemStateChangedEvent] - nest_thermostat_eca8c465_PuNbMpQGIY3hk7Nf8qJh6secrx8nUoq__set_point changed from 21.0 ℃ to 21.5 ℃
22:47:32.378 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_K changed from 294.15 K to 294.65 K
22:47:32.386 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_C changed from 21.0 ℃ to 21.5 ℃
22:47:32.387 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_F changed from 71.0 °F to 70.70 °F
22:47:32.392 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_U changed from 21.0 ℃ to 21.5 ℃
22:47:37.256 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'Thermostat_SP_K' received command 294
22:47:37.257 [INFO ] [ng.nest.handler.NestThermostatHandler] - >>> Received QuantityType command: 294 K
22:47:37.257 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_K changed from 294.65 K to 294.0 K
22:47:37.845 [INFO ] [smarthome.event.ItemStateChangedEvent] - nest_thermostat_eca8c465_PuNbMpQGIY3hk7Nf8qJh6secrx8nUoq__set_point changed from 21.5 ℃ to 21.0 ℃
22:47:37.846 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_K changed from 294.0 K to 294.15 K
22:47:37.851 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_F changed from 70.70 °F to 69.80 °F
22:47:37.851 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_C changed from 21.5 ℃ to 21.0 ℃
22:47:37.852 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_U changed from 21.5 ℃ to 21.0 ℃
```

As expected the item with the `%unit%` gets its unit from thet UnitProvider which may not match the unit of the current state:

```
openhab> smarthome:update Thermostat_SP_U "291.15 K"
Update has been sent successfully.

openhab> items list
Thermostat_SP_C (Type=NumberItem, State=21.0 ℃, Label=Set Point, Category=null)
Thermostat_SP_F (Type=NumberItem, State=69.80 °F, Label=Set Point, Category=null)
Thermostat_SP_K (Type=NumberItem, State=294.15 K, Label=Set Point, Category=null)
Thermostat_SP_U (Type=NumberItem, State=291.15 K, Label=Set Point, Category=null)
```

Then when I use HABPanel to send a command it gets Celsius as unit:

```
22:50:26.061 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'Thermostat_SP_U' received command 289
22:50:26.063 [INFO ] [ng.nest.handler.NestThermostatHandler] - >>> Received QuantityType command: 289 ℃
22:50:26.068 [INFO ] [smarthome.event.ItemStateChangedEvent] - Thermostat_SP_U changed from 291.15 K to 289.0 ℃
22:50:26.670 [WARN ] [inding.nest.handler.NestBridgeHandler] - Nest API error: Temperature C value is too high: 289.0
```

So it's definitely already a good improvement though it would be nice if items using `%unit%` could also benefit from this.

Reply:
I think we have two issues here:

1. The item did not yet receive a state, so a UI does not know which unit to send for a command. In this case the state description and the fallback to defaults from UnitProvider are the only possibilities we have.
For the "Power" example and others where a default is not yet defined the defaults map in the UnitProvider should be extended.

2. The item already has a state with a unit. A UoM capable UI should now pick up the current unit and send commands in this unit. Any other UI continues sending plain numbers (DecimalType) commands without any clue about the unit. I don't think taking the item state's unit is superior to taking the state description's unit or the unit provider default.